readme.md
![](https://github.com/zeit/art/blob/e081cf46e6609b51ac485dcc337ac6644c0da5e7/slackin/repo-banner.png)

## Features

Slackin-extended is a more customizable and extendable fork of the original [slackin](https://github.com/rauchg/slackin).

- A landing page you can point users to fill in their emails and receive an invite (`https://slack.yourdomain.com`)
- An `<iframe>` badge to embed on any website that shows connected users in *realtime* with socket.io.
- A SVG badge that works well from static mediums (like GitHub README pages)
- Abuse prevention via [Google reCAPTCHA](https://www.google.com/recaptcha/intro/), including Invisible reCAPTCHA.
- Color scheme customization on both the landing page and the badge.

Check out the [Demo](https://slackin.now.sh/) or read more about the [motivations and history](http://rauchg.com/slackin) behind the original Slackin. For the list of changes in `slackin-extended` see [HISTORY.md](HISTORY.md).

## Usage

Set up [Now](https://zeit.co/now) on your device and run this command. [Now](https://zeit.co/now) will automatically ask you for the required information and then spin up a deployment instantly. That easy!

```bash
now emedvedev/slackin-extended
```

Other platforms:

- [Heroku](https://heroku.com/deploy?template=https://github.com/emedvedev/slackin-extended/tree/master)
- [Azure](https://azuredeploy.net/)
- [OpenShift](https://github.com/rauchg/slackin/wiki/OpenShift)
- [IBM Bluemix](https://bluemix.net/deploy?repository=https://github.com/emedvedev/slackin-extended)

A Docker container is available on Docker Hub: [emedvedev/slackin-extended](https://hub.docker.com/r/emedvedev/slackin-extended/).

To host Slackin-extended yourself, install it via `npm` and launch it on your server:

```
$ npm install -g slackin-extended
$ slackin "workspace-id" "your-slack-token"
```

### Badges

#### Realtime ([demo](https://cldup.com/IaiPnDEAA6.gif))

```html
<script async defer src="https://slack.yourdomain.com/slackin.js"></script>
<!-- append "?large" to the URL for the large version -->
```

#### SVG ([demo](https://cldup.com/jWUT4QFLnq.png))

```html
<img src="https://slack.yourdomain.com/badge.svg">
```

By default, the badge background color will be set to your accent color, but you can customize both background and text colors through the query string:

```html
<img src="https://slack.yourdomain.com/badge.svg?bg=e01563&fg=ffffff">
```

### Parameters

Every CLI parameter, including mandatory arguments (workspace ID and token), can alternatively be configured through environment variables.

| Flag | Short | Environment variable | Default | Description |
| --- | --- | --- | --- | --- |
| Positional 1 | — | `SLACK_SUBDOMAIN` | **Required** | Slack workspace ID (https://**{this}**.slack.com) |
| Positional 2 | — | `SLACK_API_TOKEN` | **Required** | [API token](https://get.slack.help/hc/en-us/articles/215770388-Creating-and-regenerating-API-tokens) |
| --port | -p | `SLACKIN_PORT` | `3000` | Port to listen on |
| --hostname | -h | `SLACKIN_HOSTNAME` | `'0.0.0.0'` | Hostname to listen on |
| --channels | -c | `SLACKIN_CHANNELS` | `''` | One or more comma-separated channel names to allow single-channel guests |
| --emails | -e | `SLACKIN_EMAILS` | `''` | Restrict sign-up to a list of emails (comma-separated; wildcards are supported) |
| --interval | -i | `SLACKIN_INTERVAL` | `60000` | How frequently (ms) to poll Slack |
| --path | -P | `SLACKIN_PATH` | `/` | Path to serve slackin under |
| --silent | -s | `SLACKIN_SILENT` | `false` | Do not print out warnings or errors |
| --cors | -x | `SLACKIN_CORS` | `false` | Enable CORS for all routes |
| --analytics | -a | `SLACKIN_ANALYTICS` | `''` | Google Analytics ID |
| --recaptcha-secret | -R | `RECAPTCHA_SECRET` | `''` | reCAPTCHA secret |
| --recaptcha-sitekey | -K | `RECAPTCHA_SITEKEY` | `''` | reCAPTCHA sitekey |
| --recaptcha-invisible | -I | `RECAPTCHA_INVISIBLE` | `''` | Use [invisible reCAPTCHA](https://developers.google.com/recaptcha/docs/invisible) |
| --theme | -T | `SLACKIN_THEME` | `light` | Color scheme to use, `light` or `dark` |
| --accent | -A | `SLACKIN_ACCENT` | `#e01563` for `light`, `#9a0e44` for `dark` | Accent color to use instead of a theme default |
| --coc | -C | `SLACKIN_COC` | `''` | Full URL to a CoC that needs to be agreed to |
| --css | -S | `SLACKIN_CSS` | `''` | Full URL to a custom CSS file to use on the main page |
| | | `SLACKIN_PROXY` | `false` | Trust proxy headers (only use if Slackin is served behind a reverse proxy) |
| | | `SLACKIN_HTTPS_REDIRECT` | `''` | If a domain name is specified in this parameter and `SLACKIN_PROXY` is set to `true`, Slackin will redirect requests with `x-forwarded-proto === 'http'` to `https://<SLACKIN_HTTPS_REDIRECT>/<original URL>` |
| | | `SLACKIN_LETSENCRYPT` | `''` | [Let's Encrypt](https://letsencrypt.org/) challenge response |

## Extras

### Tips and tricks

* Please use reCAPTCHA for your Slack to avoid request flooding by spambots. Here is where to [get your secret and
sitekey](https://www.google.com/recaptcha/admin).

* [Invisible reCAPTCHA](https://developers.google.com/recaptcha/docs/invisible) is recommended for the realtime badge users, as the regular CAPTCHA challenge is broken inside the iframe that opens on badge click.

* You can find or generate your API test token at [api.slack.com/web](https://api.slack.com/web) – note that the user you use to generate the token must be an admin. You need to create a dedicated `@slackin-inviter` user (or similar), mark that user an admin, and use a test token from that dedicated admin user.  Note that test tokens have actual permissions so you do not need to create an OAuth 2 app. Also check out the Slack docs on [generating a test token](https://get.slack.help/hc/en-us/articles/215770388-Creating-and-regenerating-API-tokens).

* **Important:** If you use Slackin in single-channel mode, you'll only be
able to invite as many external accounts as paying members you have
times 5. If you are not getting invite emails, this might be the reason.
Workaround: sign up for a free org, and set up Slackin to point to it
(all channels will be visible).

### Development

1. [Fork](https://help.github.com/articles/fork-a-repo/) this repository to your own GitHub account and then [clone](https://help.github.com/articles/cloning-a-repository/) it to your local device
2. Uninstall slackin if it's already installed: `npm uninstall -g slackin-extended`
3. Link it to the global module directory: `npm link`
4. Transpile the source code and watch for changes: `npm start`

Yay! Now can use the `slackin` command everywhere.

### Extending templates and themes

If changing accent colors and switching between light/dark themes isn't enough for you, all templates are stored inside `views/` and can be easily modified. Stylesheets are stored in `scss/`, and both themes have a documented example of colors, variables and additional styles necessary to create your own.

If you end up creating another theme, you're more than welcome to submit a PR!

### API

Loading `slackin` will return a `Function` that creates a `HTTP.Server` instance:

```js
const slackin = require('slackin-extended')

slackin.default({
  token: 'SLACK-API-TOKEN', // reqired
  org: 'your-slack-subdomain', // required
  interval: 60000, // polling interval for Slack servers
  path: '/some/path/you/host/slackin/under/', // defaults to '/'
  cors: false, // set to "true" to enable CORS
  recaptcha: {
    secret: 'RECAPTCHA-SECRET', // reCAPTCHA secret
    sitekey: 'RECAPTCHA-SITEKEY', // reCAPTCHA sitekey
    invisible: true, // enable Invisible reCAPTCHA
  },
  analytics: 'GOOGLE-ANALYTICS-ID', // Google Analytics ID
  theme: 'light', // color scheme: "light" or "dark"
  accent: 'teal', // accent color (for buttons, text accents, etc.)
  css: 'https://example.org/slackin-extra.css', // external CSS file to include
  channels: 'community-public,testing', // for single-channel mode
  emails: 'john@gmail.com,jane@gmail.com,*@johnandjane.com', // email whitelist
  coc: 'https://example.org/code-of-conduct', // code of conduct link
  proxy: false, // set to "true" if Slackin is set up behind a reverse proxy
  redirectFQDN: 'slack.myorg.com', // redirect requests with x-forwarded-proto === 'http' to this domain
  letsencrypt: 'LETSENCRYPT-CHALLENGE', // Let's Encrypt challenge
  silent: false, // suppress warnings
}).listen(3000)
```

This will show response times from Slack and how many online users you have on the console. The returned `http.Server` has an `app` property that is the `express` application that you can define or override routes on.

All the metadata for your organization can be fetched via a JSON HTTP request to `/data`.


HISTORY.md
1.0.0 / 2017-10-27
==================

  * **`slackin-extended` forking point**
  * change the templating engine to Pug
  * change the CSS engine to SCSS
  * change the default Slack API polling interval to 60s
  * update the required `node` version
  * update all dependencies
  * make debug logging more explicit with the `debug` module
  * make the executable fully configurable through environment variables
  * make reCAPTCHA optional
  * add Invisible reCAPTCHA support
  * add Google Analytics support
  * add theme and accent color support for the sign-up page
  * add color customization for the SVG badge
  * add wildcard support to e-mail restrictions
  * add HTTP -> HTTPS redirect and the `trust proxy` setting
  * modify the Docker image to be compatible with Google App Engine
  * document every parameter and environment variable
  * document extending templates and themes
  * respect the `retry-after` Slack API header ([slackin/#303](https://github.com/rauchg/slackin/pull/303) by [@szkl](https://github.com/szkl))
  * fix a link to Slack API tokens in `app.json` ([slackin/#313](https://github.com/rauchg/slackin/pull/313) by [@Meandmybadself](https://github.com/Meandmybadself))
  * fix README mention of `?large` ([slackin/#269](https://github.com/rauchg/slackin/pull/269) by [@laughinghan](https://github.com/laughinghan))
  * fix reCAPTCHA rendering inside iframe ([slackin/#332](https://github.com/rauchg/slackin/pull/332) by [@toolmantim](https://github.com/toolmantim))
  * return the self-hosting instructions to README ([slackin/#336](https://github.com/rauchg/slackin/pull/336) by [@Daniel15](https://github.com/Daniel15))
  * add `dist` to `.gitignore` ([slackin/#343](https://github.com/rauchg/slackin/pull/343) by [@Ilshidur](https://github.com/Ilshidur))
  * commit `package-lock.json` ([slackin/#345](https://github.com/rauchg/slackin/pull/345) by [@Ilshidur](https://github.com/Ilshidur))
  * use hash URLs to link to a pre-selected channel ([slackin/#111](https://github.com/rauchg/slackin/pull/345) by [@mattgorecki](https://github.com/mattgorecki))

0.13.0 / 2016-07-20
===================

  * index: add `/data` endpoint to get data through a JSON API

0.12.0 / 2016-07-20
===================

  * add options cors support
  * slack: improve error handling

0.11.3 / 2016-07-19
===================

  * package: don't forget about assets

0.11.2 / 2016-07-19
===================

  * package: fix broken `main`

0.11.1 / 2016-07-19
===================

  * package: fix usecase of `require('slackin')`

0.11.0 / 2016-07-19
===================

  * index: expose express app as `app` on the returned value of `slackin`

0.10.3 / 2016-07-19
===================

  * package: only publish `bin` and `dist`

0.10.2 / 2016-07-19
===================

  * package: remove postinstall

0.10.1 / 2016-07-19
===================

  * moving babel and gulp into dev deps [@rauchg]
  * package: remove `start` as it requires ENV stuff [@rauchg]

0.10.0 / 2016-07-19
===================

  * package: bump `socket.io` for security fixes

0.9.0 / 2016-07-19
==================

  * Important bug fixes (#208) [@leo]
  * Switch to args, lint code and clean up (#200) [@leo]
  * Update docs with more clear token generation info
  * Change heroku deploy to master
  * Update Dockerfile to match Procfile
  * issue #169, add turnary assignment for client sdk channels
  * Change case on the "already invited" message
  * Update Readme.md
  * Fix TypeError when submitting w/o SLACK_CHANNELS config

0.8.3 / 2016-03-08
==================

  * fix postMessage({}, '*') causes slackin to throw [@laughinghan]
  * add IBM Bluemix Deploy badge [@kevinSuttle]
  * add `?large` to `/iframe` and `/iframe/dialog` [@laughinghan]
  * convert dialog dimensions to em/rem for adjustability #prepwork [@laughinghan]
  * fix redirecting already-registered users from iframe dialog [@laughinghan]
  * add "or sign in" link to iframe too [@laughinghan]
  * render a hidden form field for a single channel [@nickstenning]
  * add .travis.yml [@jszwedko]
  * fix test for successful invite [@jszwedko]
  * update `babel-register` module name in mocha opts [@jszwedko]
  * update readme for babel 6 change [@danreeves]
  * shorten already-invited message [@MaxWofford]
  * add success message for already signed up users [@MaxWofford]
  * fix max listeners warnings [Hardeep Shoker]

0.8.2 / 2016-01-13
==================

  * use better supported anonymous function syntax [WouterSioen]
  * bump socket.io

0.8.1 / 2016-01-06
==================

  * index: fix `import` order issue

0.8.0 / 2016-01-06
==================

  * bump socket.io to 1.4.0 for sec advisory
  * add deploy to Azure Button
  * build: Update builds to use gulp and babel6 (Fixes #122)
  * adds instructions for deploying to Cloud Foundry
  * redirect signed up users to Slack
  * fix --path option parsing

0.7.3 / 2015-11-29
==================

  * package: bump `socket.io` to fix build issue

0.7.2 / 2015-11-29
==================

  * fix rendering of badge in non-chrome

0.7.1 / 2015-11-19
==================

  * fix inline badge styles

0.7.0 / 2015-11-19
==================

  * badge: make badge work for users with lots of users
  * splash: fix hover color – thank goodness
  * relative path hosting support
  * fix tests
  * add support for codes of conduct

0.6.0 / 2015-10-05
==================

  * use faster API methods: users.list to get presence
  * fix polling interval increase
  * allow hosting slackin under a path (eg: /slack)
  * fix for deployment instructions
  * add already registered link
  * use SLACK_CHANNELS instead of SLACK_CHANNEL on Heroku
  * standardize use of example host
  * allow hosting at a subfolder

0.5.1 / 2015-06-23
==================

  * added safety check for channels.
  * fix counting deleted users

0.5.0 / 2015-05-07
==================

  * allow for multiple guest channels [afeld]
  * add important notes to README
  * better error handling and tests [afeld]

lib/index.js
// their code
import express from 'express';
import sockets from 'socket.io';
import { json } from 'body-parser';
import { Server as http } from 'http';
import { readFileSync as read } from 'fs';
import remail from 'email-regex';
import cors from 'cors';
import request from 'superagent';
import dbg from 'debug';
import tinycolor from 'tinycolor2';
import match from 'micromatch';

// our code
import Slack from './slack';
import invite from './slack-invite';
import themes from './themes';

const mainLog = dbg('slackin:main');
const inviteLog = dbg('slackin:invite');
const slackLog = dbg('slackin:slack');

export default function slackin({
  token,
  org,
  path = '/',
  interval = 60000,
  cors: useCors = false,
  recaptcha = {},
  analytics,
  theme: themeID,
  accent,
  css,
  channels: channelFilter,
  emails,
  coc,
  proxy,
  redirectFQDN,
  letsencrypt,
  silent,
}) {
  // must haves
  if (!token) throw new Error('Must provide a `token`.');
  if (!org) throw new Error('Must provide an `org`.');
  if (
    !!(recaptcha.secret || recaptcha.sitekey || recaptcha.invisible) !==
    !!(recaptcha.secret && recaptcha.sitekey)
  ) {
    throw new Error('Both `recaptcha-secret` and `recaptcha-sitekey` must be defined to enable reCAPTCHA.');
  }

  if (!silent) {
    inviteLog.enabled = true;
    slackLog.enabled = true;
    mainLog.enabled = true;
  }

  let channels;
  if (channelFilter) {
    channels = channels.split(',').map(channel => (
      channel[0] === '#' ? channel.substr(1) : channel
    ));
  }

  let acceptedEmails;
  if (emails) {
    acceptedEmails = emailFilter.split(',');
  }

  let theme;
  if (themeID) {
    if (themeID in themes) {
      theme = themes[themeID];
    } else {
      mainLog(`Specified theme (${themeID}) not found, falling back to default`);
    }
  } else {
    theme = themes.default;
  }
  mainLog(`Theme: ${theme.name}`);
  if (accent) {
    theme.accent = tinycolor(accent).toHexString();
    mainLog(`Using a custom theme accent: ${accent}`);
  }
  theme.accentDark = tinycolor(theme.accent).darken(10).toHexString();

  // setup
  const app = express();
  const srv = http(app);
  srv.app = app;

  app.set('view engine', 'pug');

  if (useCors) {
    app.options('*', cors());
    app.use(cors());
  }

  if (proxy) {
    app.enable('trust proxy');
    if (redirectFQDN) {
      app.use((req, res, next) => {
        if (req.headers['x-forwarded-proto'] === 'http') {
          res.redirect(`https://${redirectFQDN}${req.url}`);
        } else {
          next();
        }
      });
    }
  }

  // static files
  const assets = `${__dirname}/assets`;
  app.use('/assets', express.static(assets));
  app.use('/slackin.js', express.static(`${assets}/badge.js`));

  if (letsencrypt) {
    app.get('/.well-known/acme-challenge/:id', (req, res) => {
      res.send(letsencrypt);
    });
  }

  // fetch data
  mainLog('Establishing connection with Slack');
  const slack = new Slack({
    token, interval, org, logger: slackLog,
  });
  slack.setMaxListeners(Infinity);

  // middleware for waiting for slack
  app.use((req, res, next) => {
    if (slack.ready) return next();
    return slack.once('ready', next);
  });

  app.get('/', (req, res) => {
    const { name, logo } = slack.org;
    const { active, total } = slack.users;
    if (!name) return res.send(404);

    return res
      .type('html')
      .render('main', {
        coc,
        path,
        name,
        org,
        logo,
        active,
        total,
        recaptcha,
        css,
        analytics,
        channels,
        theme,
      });
  });

  app.get('/data', (req, res) => {
    const { name, logo } = slack.org;
    const { active, total } = slack.users;
    res.send({
      name,
      org,
      coc,
      logo,
      active,
      total,
      channels,
    });
  });

  // invite endpoint
  app.post('/invite', json(), (req, res) => {
    const { channel, email } = req.body;
    const captchaResponse = req.body['g-recaptcha-response'];

    let errorMessage = null;
    if (channels && !channels.includes(channel)) {
      errorMessage = 'Not a permitted channel';
    } else if (channels && !slack.getChannelId(channel)) {
      errorMessage = `Channel "${channel}" not found`;
    } else if (!email) {
      errorMessage = 'No email provided';
    } else if (recaptcha.secret && (!captchaResponse || !captchaResponse.length)) {
      errorMessage = 'Invalid captcha';
    } else if (!remail().test(email)) {
      errorMessage = 'Invalid email';
    } else if (emails && !match.any(email, acceptedEmails)) {
      errorMessage = 'Your email is not on the accepted list.';
    } else if (coc && req.body.coc !== '1') {
      errorMessage = 'Agreement to CoC is mandatory';
    }
    if (errorMessage) {
      return res.status(400).json({ msg: errorMessage });
    }

    const captchaData = {
      secret: recaptcha.secret,
      response: captchaResponse,
      remoteip: req.connection.remoteAddress,
    };

    const captchaCallback = (err) => {
      if (err) return res.status(400).send({ msg: err });

      return invite({
        token, org, email, logger: inviteLog, channel: slack.channel ? slack.channel.id : null,
      }, (inviteErr) => {
        if (inviteErr) {
          if (inviteErr.message === 'Sending you to Slack...') {
            return res.status(303).json({ msg: inviteErr.message, redirectUrl: `https://${org}.slack.com` });
          }
          return res.status(400).json({ msg: inviteErr.message });
        }
        return res.status(200).json({ msg: 'WOOT. Check your email!' });
      });
    };

    if (recaptcha.secret) {
      return request.post('https://www.google.com/recaptcha/api/siteverify')
        .type('form')
        .send(captchaData)
        .end(captchaCallback);
    }

    return captchaCallback();
  });

  // iframe
  app.get('/iframe', (req, res) => {
    const logo = read(`${__dirname}/../assets/slack.svg`).toString('base64');
    const js = read(`${__dirname}/../assets/iframe.js`).toString();
    const extraCss = read(`${__dirname}/../assets/iframe-button.css`).toString();
    const large = 'large' in req.query;
    const { active, total } = slack.users;
    res.type('html');
    res.render('iframe', {
      path, active, total, large, logo, js, extraCss, css,
    });
  });

  app.get('/iframe/dialog', (req, res) => {
    const large = 'large' in req.query;
    const { name } = slack.org;
    const { active, total } = slack.users;
    if (!name) {
      res.sendStatus(404);
      return;
    }
    res.type('html');
    res.render('main', {
      coc,
      path,
      name,
      org,
      active,
      total,
      large,
      recaptcha,
      analytics,
      channels,
      theme,
      iframe: true,
    });
  });

  // badge rendering
  app.get('/badge.svg', (req, res) => {
    res.type('svg');
    res.set('Cache-Control', 'max-age=0, no-cache');
    res.set('Pragma', 'no-cache');
    const options = {
      total: slack.users.total,
      active: slack.users.active,
      bg: req.query.bg ? tinycolor(req.query.bg).toHexString() : theme.accent,
    };
    if (req.query.fg) {
      options.fg = tinycolor(req.query.fg).toHexString();
    } else {
      options.fg = tinycolor(options.bg).isDark() ? '#fff' : '#333';
    }
    res.render('badge-svg', options);
  });

  // realtime
  sockets(srv).on('connection', (socket) => {
    const change = (key, val) => socket.emit(key, val);
    slack.on('change', change);
    socket.emit('data', slack.users);
    socket.on('disconnect', () => {
      slack.removeListener('change', change);
    });
  });

  return srv;
}
